File: python/examples/tools/basic.py
import math

from mirascope import llm


@llm.tool
def sqrt_tool(number: float) -> float:
    """Computes the square root of a number"""
    return math.sqrt(number)


@llm.call("openai/gpt-5-mini", tools=[sqrt_tool])
def math_assistant(query: str):
    return query


response = math_assistant("What's the square root of 4242?")
tool_call = response.tool_calls[0]
print(tool_call)
# ToolCall(type='tool_call', id='...', name='sqrt_tool', args='{"number":4242}')

# Looks up sqrt_tool by name and calls sqrt_tool.execute(tool_call)
tool_output = response.toolkit.execute(tool_call)
print(tool_output)
# ToolOutput(type='tool_output', id='...', name='sqrt_tool', value=65.13063795173512)

answer = response.resume(tool_output)
print(answer.pretty())
# The square root of 4242 is approximately 65.13063795173512.

File: typescript/examples/tools/basic.ts
import { llm } from "mirascope";

const sqrtTool = llm.defineTool<{ number: number }>({
  name: "sqrt_tool",
  description: "Computes the square root of a number",
  fieldDefinitions: {
    number: "The number to compute the square root of",
  },
  tool: ({ number }) => Math.sqrt(number),
});

const mathAssistant = llm.defineCall<{ query: string }>({
  model: "openai/gpt-4o-mini",
  tools: [sqrtTool],
  template: ({ query }) => query,
});

const response = await mathAssistant({
  query: "What's the square root of 4242?",
});

const toolCall = response.toolCalls[0]!;
console.log(toolCall);
// ToolCall { type: 'tool_call', id: '...', name: 'sqrt_tool', args: { number: 4242 } }

// Execute all tool calls and get outputs
const toolOutputs = await response.executeTools();
console.log(toolOutputs[0]);
// ToolOutput { type: 'tool_output', id: '...', name: 'sqrt_tool', output: 65.13063795173512, error: null }

const answer = await response.resume(toolOutputs);
console.log(answer.text());
// The square root of 4242 is approximately 65.13063795173512.
